Mathlib/LinearAlgebra/TensorProduct/Submodule.lean

```diff
@@ -0,0 +1,202 @@
+def mulMap : M ⊗[R] N →ₗ[R] S :=
+  TensorProduct.lift <| ((LinearMap.domRestrict' N).comp <| .mul R S).domRestrict M
+
+@[simp]
+theorem mulMap_tmul (m : M) (n : N) : mulMap M N (m ⊗ₜ[R] n) = m.1 * n.1 := rfl
+
+theorem mulMap_op :
+    mulMap (equivOpposite.symm (MulOpposite.op M)) (equivOpposite.symm (MulOpposite.op N)) =
+    (MulOpposite.opLinearEquiv R).toLinearMap ∘ₗ mulMap N M ∘ₗ
+    (TensorProduct.congr
+      (LinearEquiv.ofSubmodule' (MulOpposite.opLinearEquiv R).symm M)
+      (LinearEquiv.ofSubmodule' (MulOpposite.opLinearEquiv R).symm N) ≪≫ₗ
+    TensorProduct.comm R M N).toLinearMap :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comm_of_commute (hc : ∀ (m : M) (n : N), Commute m.1 n.1) :
+    mulMap N M = mulMap M N ∘ₗ TensorProduct.comm R N M := by
+  refine TensorProduct.ext' fun n m ↦ ?_
+  simp_rw [LinearMap.comp_apply, LinearEquiv.coe_coe, TensorProduct.comm_tmul, mulMap_tmul]
+  exact (hc m n).symm
+
+theorem mulMap_comp_rTensor (M' : Submodule R S) (hM : M' ≤ M) :
+    mulMap M N ∘ₗ (inclusion hM).rTensor N = mulMap M' N :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comp_lTensor (N' : Submodule R S) (hN : N' ≤ N) :
+    mulMap M N ∘ₗ (inclusion hN).lTensor M = mulMap M N' :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_comp_map_inclusion (M' N' : Submodule R S) (hM : M' ≤ M) (hN : N' ≤ N) :
+    mulMap M N ∘ₗ TensorProduct.map (inclusion hM) (inclusion hN) = mulMap M' N' :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_eq_mul'_comp_mapIncl : mulMap M N = .mul' R S ∘ₗ TensorProduct.mapIncl M N :=
+  TensorProduct.ext' fun _ _ ↦ rfl
+
+theorem mulMap_range : LinearMap.range (mulMap M N) = M * N := by
+  refine le_antisymm ?_ (mul_le.2 fun m hm n hn ↦ ⟨⟨m, hm⟩ ⊗ₜ[R] ⟨n, hn⟩, rfl⟩)
+  rintro _ ⟨x, rfl⟩
+  induction x using TensorProduct.induction_on with
+  | zero => rw [_root_.map_zero]; exact zero_mem _
+  | tmul a b => exact mul_mem_mul a.2 b.2
+  | add a b ha hb => rw [_root_.map_add]; exact add_mem ha hb
+
+/-- If `M` and `N` are submodules in an algebra `S` over `R`, there is the natural map
+`M ⊗[R] N →ₗ[R] M * N` induced by multiplication in `S`,
+which is surjective (`Submodule.mulMap'_surjective`). -/
+def mulMap' : M ⊗[R] N →ₗ[R] ↥(M * N) :=
+  (LinearEquiv.ofEq _ _ (mulMap_range M N)).toLinearMap ∘ₗ (mulMap M N).rangeRestrict
+
+@[simp]
+theorem val_mulMap'_tmul (m : M) (n : N) : (mulMap' M N (m ⊗ₜ[R] n) : S) = m.1 * n.1 := rfl
+
+theorem mulMap'_surjective : Function.Surjective (mulMap' M N) := by
+  simp_rw [mulMap', LinearMap.coe_comp, LinearEquiv.coe_coe, EquivLike.comp_surjective,
+    LinearMap.surjective_rangeRestrict]
+
+/-- If `N` is a submodule in an algebra `S` over `R`, there is the natural map
+`i(R) ⊗[R] N →ₗ[R] N` induced by multiplication in `S`, here `i : R → S` is the structure map.
+This is promoted to an isomorphism as `Submodule.lTensorOne`. Use that instead. -/
+def lTensorOne' : (1 : Submodule R S) ⊗[R] N →ₗ[R] N :=
```

Comment:
I don't think so; here I emphasis that it should be a `Submodule`. This map is used to prove that `(1 : Submodule R S)` is linearly disjoint to any other submodule.

By the way, for subalgebra I think it should be written as `(⊥ : Subalgebra R S)`.

Reply:
This is coerced into a type so they are definitionally the same as long as the underlying subtype is?
And I think one and bot are the same thing for subalgebras?

Reply:
No, there is no `(1 : Subalgebra R S)`.

Reply:
I'm sorry, but my opinion is keep using `(1 : Submodule R S)` since this file is for submodule, not subalgebra. I'll mark this conversation as resolved. Feel free to unresolve it if there are further comments.